internal/provider/data_host.go

```diff
@@ -0,0 +1,434 @@
+				Computed:    true,
+				Description: "The memory details of the ESXi host.",
+				Attributes: map[string]schema.Attribute{
+					"total_capacity_mb": schema.Float64Attribute{
+						Computed: true,
+					},
+					"used_capacity_mb": schema.Float64Attribute{
+						Computed: true,
+					},
+				},
+			},
+			"storage": schema.ListNestedAttribute{
+				Computed:    true,
+				Description: "The storage details of the ESXi host.",
+				NestedObject: schema.NestedAttributeObject{
+					Attributes: map[string]schema.Attribute{
+						"total_capacity_mb": schema.Float64Attribute{
+							Computed: true,
+						},
+						"used_capacity_mb": schema.Float64Attribute{
+							Computed: true,
+						},
+						"disks": schema.ListNestedAttribute{
+							Computed:    true,
+							Description: "The disks of the ESXi host.",
+							NestedObject: schema.NestedAttributeObject{
+								Attributes: map[string]schema.Attribute{
+									"capacity_mb": schema.Float64Attribute{
+										Computed: true,
+									},
+									"disk_type": schema.StringAttribute{
+										Computed: true,
+									},
+									"manufacturer": schema.StringAttribute{
+										Computed: true,
+									},
+									"model": schema.StringAttribute{
+										Computed: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			"physical_nics": schema.ListNestedAttribute{
+				Computed:    true,
+				Description: "The physical NICs of the ESXi host.",
+				NestedObject: schema.NestedAttributeObject{
+					Attributes: map[string]schema.Attribute{
+						"device_name": schema.StringAttribute{
+							Computed: true,
+						},
+						"mac_address": schema.StringAttribute{
+							Computed: true,
+						},
+						"speed": schema.Int64Attribute{
+							Computed: true,
+						},
+						"unit": schema.StringAttribute{
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (d *DataSourceHost) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data []HostModel
+
+	params := hosts.NewGetHostsParamsWithContext(ctx).
+		WithTimeout(constants.DefaultVcfApiCallTimeout)
+
+	hostPayload, err := d.client.Hosts.GetHosts(params)
+	if err != nil {
+		resp.Diagnostics.Append(diag.NewErrorDiagnostic("Failed to retrieve hosts", err.Error()))
+		return
+	}
+
+	for _, element := range hostPayload.Payload.Elements {
+		// Debugging logs to print the values of id and status
+		tflog.Debug(ctx, "Mapping host", map[string]interface{}{
+			"fqdn":   element.Fqdn,
+			"id":     element.ID,
+			"status": element.Status,
+		})
+
+		host := mapHostElementToModel(element)
+		data = append(data, host)
+	}
+
+	// Debugging log to trace the mapped data
+	tflog.Debug(ctx, "Mapped data", map[string]interface{}{
+		"hosts": data,
+	})
+
+	resp.State.Set(ctx, &data)
+}
+
+func mapHostElementToModel(element *models.Host) HostModel {
+	// Inline dereferencing logic
+	domainID := derefString(element.Domain.ID)
```

Comment:
The "framework" provides other means of handling nil values.
It can differentiate unset (unknown) and nil values from actual valid ones.

Instead of having `derefString` which defaults to an empty string you can use `types.StringNull()` or `types.StringUnknown()`.

I haven't tried this myself, but it looks like it can convey more information than simply masking nils as empty strings